File: AlgorithmsCsharp/Algorithms/SortingAlgorithms/QuickSort.cs
namespace Algorithms.SortingAlgorithms;

public static class QuickSort
{
    public static void Sort(int[] inputArray)
    {
        SortInternal(inputArray, 0, inputArray.Length - 1);
    }

    private static int[] SortInternal(int[] array, int minIndex, int maxIndex)
    {
        if (minIndex >= maxIndex)
        {
            return array;
        }

        var pivotIndex = Partition(array, minIndex, maxIndex);
        SortInternal(array, minIndex, pivotIndex - 1);
        SortInternal(array, pivotIndex + 1, maxIndex);

        return array;
    }

    private static void Swap(ref int x, ref int y)
    {
        var t = x;
        x = y;
        y = t;
    }

    /// <summary>
    /// Divides the array into two subarrays, one with elements smaller than the pivot element, and one with elements
    /// </summary>
    /// <returns>The index of the pivot element.</returns>
    private static int Partition(int[] array, int minIndex, int maxIndex)
    {
        var pivot = minIndex - 1;
        for (var i = minIndex; i < maxIndex; i++)
        {
            if (array[i] < array[maxIndex])
            {
                pivot++;
                Swap(ref array[pivot], ref array[i]);
            }
        }

        pivot++;
        Swap(ref array[pivot], ref array[maxIndex]);
        return pivot;
    }
}

File: AlgorithmsCPlusPlus/Sorting/QuickSort.cpp
#include <iostream>
#include "QuickSort.h"


void QuickSort::sort(int arr[], int size)
{
    quickSort(arr, 0, size - 1);
}

int QuickSort::partition(int array[], int low, int high)
{   
  int pivot = array[high];
  
  int i = (low - 1);

  for (int j = low; j < high; j++) {
    if (array[j] <= pivot) {
        
      i++;
      
      swap(&array[i], &array[j]);
    }
  }
  
  swap(&array[i + 1], &array[high]);
  
  return (i + 1);
}
 
 void QuickSort::quickSort(int arr[], int start, int end)
{  
    if(start < end)
    {   
        int partitionIndex = partition(arr, start, end);
        quickSort(arr, start, partitionIndex - 1); 
        quickSort(arr, partitionIndex + 1, end); 
    }
}

void QuickSort::swap(int* a, int* b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}